[14.0] Add delivery_carrier_package_measure_required

TDu opened the pull request.

TDu committed:
Add delivery_carrier_package_measure_required

This module adds the possibility to flag some measurements on package to be
required for delivery.

rousseldenis commented:
Hi @TDu, thanks for this.

As it is quite linked to delivery carrier, I would have named the module with another name, mentioning at least delivery carrier (the flow is not available on all pickings).

TDu commented:
> Hi @TDu, thanks for this.
> 
> As it is quite linked to delivery carrier, I would have named the module with another name, mentioning at least delivery carrier (the flow is not available on all pickings).

Totally agree. I changed it. I hope for the better.

simahawk commented:
I'd go for `delivery_carrier_package_measure_required` or `delivery_carrier_require_package_measure`. Both are shorter :wink:

simahawk commented on delivery_carrier_required_package_measurement/wizard/choose_delivery_package.py:
```diff
@@ -0,0 +1,35 @@
+    package_width = fields.Integer()
+    dimension_uom_id = fields.Many2one(
+        "uom.uom",
+        default=_default_dimension_uom_id,
```
```suggestion
        default=lambda self: self._default_dimension_uom_id(),
```
Then you can move the method below.

TDu replied:
Changed

simahawk reviewed (changes requested):
LG overall, I'd change the name as proposed.

TDu commented:
Thanks for the reviews, I changed the name of the module.

simahawk commented:
@rousseldenis fine for you?

rousseldenis commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,34 @@
+    def button_validate(self):
+        for pick in self:
+            if not pick._check_required_package_measurement():
+                raise UserError(
```
Could you just add a test for this case as this is the aim of the module, no ?

TDu replied:
Sure, good  idea. I just did.

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,34 @@
+                return False
+            if carrier.package_width_required and not package.width:
+                return False
+        return True
```
Why no check on package_weight_required ?

TDu replied:
Probably forgot :shrug: 
But now it is added.

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,34 @@
+            if not pick._check_required_package_measurement():
+                raise UserError(
+                    _(
+                        "Some destination package do not have the required measurement set."
```
```suggestion
                        "A destination package is lacking the required measurement."
```

TDu replied:
Changed

jbaudoux commented on delivery_carrier_package_measure_required/models/delivery_carrier.py:
```diff
@@ -0,0 +1,13 @@
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    package_height_required = fields.Boolean(string="Height required")
```
A carrier can transport different delivery packages requiring different measurement.
Like a pallet only needs height, a parcel needs length,width,height,weight
I would move this from `delivery.carrier` to `product.packaging` (delivery package).

jbaudoux replied:
And then to be consistent, that would maybe mean to rename module to `delivery_package_measure_required` :-/

TDu replied:
I moved the fields to `product.packaging` but did not rename the module

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,34 @@
+        carrier = self.carrier_id
+        packages = self.move_line_ids.mapped("result_package_id")
+        for package in packages:
+            if carrier.package_height_required and not package.height:
```
and then here check `package.packaging_id` instead of `carrier`

TDu replied:
Done

jbaudoux commented on delivery_carrier_package_measure_required/views/delivery_carrier.xml:
```diff
@@ -0,0 +1,25 @@
+                            <field name="package_height_required" />
+                            <field name="package_length_required" />
+                            <field name="package_weight_required" />
+                            <field name="package_width_required" />
```
Can you change order to:
* Length
* Width
* Height
* Weight

Like here:
![image](https://user-images.githubusercontent.com/7802725/193588727-c9a4f913-9e94-4010-9862-43788eff755b.png)

TDu replied:
Done

jbaudoux commented on delivery_carrier_package_measure_required/wizard/choose_delivery_package.xml:
```diff
@@ -0,0 +1,54 @@
+        <field name="model">choose.delivery.package</field>
+        <field name="inherit_id" ref="delivery.choose_delivery_package_view_form" />
+        <field name="arch" type="xml">
+          <xpath expr="//div[@name='package_weight']" position="after">
```
Can you change order to: length,width,height,weight ?

TDu replied:
Yes I did.

jbaudoux reviewed (changes requested):
A more granular configuration per delivery package would be more suitable

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,34 @@
+
+    def _check_required_package_measurement(self):
+        self.ensure_one()
+        if not self.carrier_id:
```
This needs to use `ship_carrier_id` from `stock_picking_delivery_link` module as the carrier is defined only on the ship step and here we likely are on a pack step in the chain of pick-pack-ship

TDu replied:
@jbaudoux If we move the required measurement flag on to the `product.packaging` model there is no need to check anything on the carrier. 
Would that be coherent for you ?

jbaudoux replied:
Yes, that's correct. It's easier. Only the delivery package will tell what you need as measurement.

TDu replied:
Done

jbaudoux reviewed (approved):
Tested. Works perfectly.

Can you squash the commits for merging?

TDu commented:
Squashed and ready to merge ?
@simahawk @rousseldenis

lmignon commented on delivery_carrier_package_measure_required/wizard/choose_delivery_package.py:
```diff
@@ -0,0 +1,35 @@
+    )
+    package_width_required = fields.Boolean(
+        related="delivery_packaging_id.package_width_required"
+    )
```
IMO you should add an `api.contrains` method to validate that the required dimensions are filled on the backend side...

TDu replied:
You are right. I changed the implementation to use an `api.constrains`

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_quant_package.py:
```diff
@@ -0,0 +1,36 @@
+
+    @api.constrains("length", "width", "height", "weight")
+    # The boolean field use to check if a dimension is required are intentionally left out.
+    # To not raise error when changingt packaging configuration.
```
```suggestion
    # To not raise error when changing packaging configuration.
```

jbaudoux commented:
@lmignon Can you update your review?

lmignon commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,26 @@
+        If package is None all result packages on the stock picking are check
+        otherwise only the one.
+        """
+        self.ensure_one()
```
@TDu Why do you limit the call to only one picking? The only change to support multi records would be `packages = self.mapped('move_line_ids.result_package_id')`

lmignon commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,26 @@
+        if package:
+            packages = packages & package
+        for package in packages:
+            package._check_required_dimension()
```
The `_check_required_dimension` method is designed to support mutti records. The loop is therefore useless: `packages._check_required_dimension()`

lmignon commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,26 @@
+            package._check_required_dimension()
+
+    def button_validate(self):
+        for pick in self:
```
By supporting multi records, your code is simpler: `self._check_required_package_measurement()`

lmignon commented on delivery_carrier_package_measure_required/models/stock_quant_package.py:
```diff
@@ -0,0 +1,36 @@
+                raise ValidationError(
+                    _(
+                        "The measurement(s) [{}] are required "
+                        "on the package and need to be set."
```
You should add the package name into the message.

TDu commented:
@lmignon  Thank you for your reviews. All very good points that I just implemented in the last fixup.

lmignon commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,24 @@
+        packages = self.move_line_ids.mapped("result_package_id")
+        if package:
+            packages = packages & package
+        for package in packages:
+            package._check_required_dimension()
```
@TDu You missed this one....
```suggestion
        packages._check_required_dimension()
```

TDu replied:
Oh dear :facepalm:

simahawk commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,23 @@
+    def _check_required_package_measurement(self, package=None):
+        """Check the required measurement on destination pacakges are set.
+
+        If package is None all result packages on the stock picking are check
```
hmm all packages are validated anyway.... outdated docstring or wrong code?

TDu replied:
There is a typo in docstring `check` -> `checked` but the code does what is said IMO.
From the doc on recordset operation:

> set1 & set2 returns the intersection of two recordsets, a new recordset containing only records present in both sources

simahawk replied:
Ok. so basically you are validating that a package belongs to the same picking and, if not, it will fail silently.

Then, I wonder if we really need `package=` to be there since if you want to validate a package you can simply call `package._check_required_dimension()`.
Thus, I'd say that if we keep it on the picking, the package must be validated. Eg:
```
if package and not packages:
    raise exceptions.UserError(_("Picking %s does not contain the package %"))
```
Do you see my point?

TDu commented:
I fixed the typo in comments and docstring and squashed the commits

simahawk commented on delivery_carrier_package_measure_required/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,9 @@
+This module adds the possibility to flag some measurements on package to be required for delivery.
+
+It depends on the `stock_quant_package_dimension` which adds some dimension on pacakges.
```
```suggestion
It depends on the `stock_quant_package_dimension` which adds some dimension on packages.
```

TDu replied:
fixed

simahawk commented on delivery_carrier_package_measure_required/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,9 @@
+
+It depends on the `stock_quant_package_dimension` which adds some dimension on pacakges.
+
+The required measurement can be flaged on the product packaging.
```
```suggestion
The required measurement can be flagged on the product packaging.
```

TDu replied:
fixed

TDu commented:
@simahawk  Yes now I understand, so I removed the package argument which simplify the code even more. Also added a few lines of test to improve coverage.

simahawk commented:
/ocabot merge nobump
